xarray/core/datatree.py

```diff
@@ -446,7 +440,7 @@ def ds(self) -> DatasetView:
         return DatasetView._from_node(self)
 
```

Comment:
This is really a comment for L430:

```
def ds(self) -> DatasetView:
```

This makes sense (because of wanting an immutable version of the `Dataset`), but it's causing the bulk of the remaining `mypy` issues in the tests, where `Dataset` objects are being directly assigned to `DataTree.ds`. That said, I think I'm a little surprised this is causing an issue, given the signature below for the `@ds.setter`. I'd love some guidance on resolving those `mypy` errors!

Reply:
I'm also surprised that causes a problem... Is this pattern valid typing in general for properties?

```python
class A:
    ...


class B(A):
    ...


class C:
    @property
    def foo(self) -> B:
        ...

    @property.setter
    def foo(self, value: A):
        ...
```

Reply:
that's fine until you set C.foo with A.

``` python
class A:
    ...

class B(A):
    ...

class C:
    @property
    def foo(self) -> B:
        return B()

    @foo.setter
    def foo(self, value: A):
        pass


nope = C()
nope.foo = A()
```

output:

``` text
> mypy demo/demo7.py
demo/demo7.py:20: error: Incompatible types in assignment (expression has type "A", variable has type "B")  [assignment]
```

And I think this is the open issue for fixing/ignoring this issue
https://github.com/python/mypy/issues/3004

But it seems like it's not high priority since it's been open for 7 years.